Enable static publishing

abliss opened the pull request.

abliss committed:
Freshen for modern vagrant-spk and debian distro.

Also replaces 'unlink' with 'rm -rf' to make it easier to reprovision.

abliss committed:
Add static publishing.

This merges in some of the technology of paulproteus's fork, but in a less
intrusive manner. Pushes to the `gw-pages` branch are automatically published to
the grain's public web url.

dwrensha commented on .sandstorm/usr/share/gitweb/indextext.html:
```diff
@@ -4,6 +4,12 @@
+<p>
+  <form action="publish.cgi">
+    You can also choose to publish the master branch as
+    static files, by pushing this button: <button>Publish</button>
```
Could we also add a way to unpublish? Like, it could unlink the post-receive-hook and clear out the `/var/www` directory. I'm worried about accidentally publishing things that were meant to be private. (I understand the "public ID" would also need to leak for this to actually be a problem, but "public ID" sounds like something that could easily leak.)

dwrensha commented on post-receive-hook.sh:
```diff
@@ -0,0 +1,11 @@
+set -e
+while read oldrev newrev refname
+do
+  if [ "$refname" = "refs/heads/master" ]
```
Would it possibly make more sense for the publish branch to be named `gw-pages`?  That would complete the analogy with github pages, and it would make it a bit more difficult to accidentally publish something that you mean to keep private.

dwrensha replied:
Hm... if the ref-to-publish is named `gw-pages`, then maybe we don't need a "publish" button at all.

abliss commented:
Sounds good to me; I'll respin the change. Thanks for your attention!

dwrensha commented:
If we go we the `gw-pages` branch idea, then we should make sure to explain how that works somewhere. Maybe the place where the "publish" button was would have a few sentences of explanation.

abliss commented:
Ok, now it auto-publishes on pushes to gw-pages. I added explanation to this effect on the front page. I also made it possible to "unpublish" by pushing an empty tree to the gw-pages branch (Previously, pushing a totally empty tree wouldn't actually unpublish, since `git checkout -- . ` wouldn't match any valid paths, so the hook would barf.)

I also added a force-clean step, so that the hook will explicitly clear out any stray files that may have somehow wound up in /var/www.

I decided to keep the publish button, so that the user can manually trigger a publish (in case the automatic one from the hook somehow failed) from the web browser without opening a terminal to run a git command. (This also provides a convenient way to display the public grain id.) Let me know if you're really opposed to having the button there. The replacement would be to somehow save the public id after the first time it was generated, and somehow splice that id into the indextext.html (currently it's a static page snippet that the gitweb cgi includes; I'm not sure how to cleanly inject new cgi code along that path to dynamically fetch the public grain ID. I'm also not clear on exactly how/when the public grain ID might change and invalidate our stored version.)

dwrensha commented on .sandstorm/usr/share/gitweb/publish.cgi:
```diff
@@ -0,0 +1,13 @@
+echo '(re)publishing the gw-pages branch...'
+exec 2>&1
+cd /var/repo.git
+echo 'aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa' \
```
What do all these `a`s do? Can you add a comment explaining them?

abliss replied:
Good question. I added a comment. Let me know whether it's clear enough.

dwrensha commented:
Looks good to me. Thanks!
